### PR TITLE
Some features from the PR #197 (PR splitted)

### PR DESCRIPTION
**Pull request by pitbulk**
* Comment the code.
* Remove spaces and format some lines.
* Remove unnecessary errors method.
* Improve format_cert and format_private_key.
* Fix xpath injection on xml_security.rb

### CODE REVIEW

**Comment by pitbulk** [comment]
@luisvm @daniel-g @Lordnibbler can I merge it?

File: lib/onelogin/ruby-saml/response.rb [hunk 1]
```diff
@@ -8,40 +8,58 @@
...
       PROTOCOL  = "urn:oasis:names:tc:SAML:2.0:protocol"
       DSIG      = "http://www.w3.org/2000/09/xmldsig#"
 
-      # TODO: This should probably be ctor initialized too... WDYT?
+      # TODO: Settings should probably be initialized too... WDYT?
+
+      # OneLogin::RubySaml::Settings Toolkit settings
       attr_accessor :settings
+
+      # Array with the causes
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/response.rb]: will this contain an array of Strings/Objects?

File: lib/onelogin/ruby-saml/response.rb [hunk 2]
```diff
@@ -184,6 +251,11 @@ def validate_response_state(soft = true)
         true
       end
 
+      # Extracts the first appearance that matchs the subelt (pattern)
+      # Search on any Assertion that is signed, or has a Response parent signed
+      # @param subelt [String] The XPath pattern
+      # @return [REXML::Element | nil] If any matchs, return the Element
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/response.rb]: typo `matches`

File: lib/onelogin/ruby-saml/saml_message.rb [hunk 3]
```diff
@@ -58,32 +70,53 @@ def decode_raw_saml(saml)
...
+      # @return [String] The deflated and encoded SAML Message (encoded if the compression is requested)
+      #
       def encode_raw_saml(saml, settings)
         saml = deflate(saml) if settings.compress_request
 
         CGI.escape(Base64.encode64(saml))
       end
 
+      # Base 64 decode method
+      # @param saml [String] The string message
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: `@param` should describe `encoded` not `saml`

File: lib/onelogin/ruby-saml/saml_message.rb [hunk 4]
```diff
@@ -58,32 +70,53 @@ def decode_raw_saml(saml)
...
+      # @return [String] The decoded string
+      #
       def decode(encoded)
         Base64.decode64(encoded)
       end
 
+      # Base 64 encode method
+      # @param saml [String] The string
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: same as above

File: lib/onelogin/ruby-saml/saml_message.rb [hunk 5]
```diff
@@ -58,32 +70,53 @@ def decode_raw_saml(saml)
...
-      #
       # @param string [String] string to check the encoding of
       # @return [true, false] whether or not the string is base64 encoded
+      #
       def base64_encoded?(string)
         !!string.gsub(/[\r\n]|\\r|\\n/, "").match(BASE64_FORMAT)
       end
 
+      # Inflate method
+      # @param saml [String] The string
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: `@param deflated ...` instead

File: lib/onelogin/ruby-saml/saml_message.rb [hunk 6]
```diff
@@ -58,32 +70,53 @@ def decode_raw_saml(saml)
...
+      # @return [String] The inflated string
+      #
       def inflate(deflated)
         Zlib::Inflate.new(-Zlib::MAX_WBITS).inflate(deflated)
       end
 
+      # Deflate method
+      # @param saml [String] The string
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: `@param inflated ...`

File: lib/onelogin/ruby-saml/settings.rb [hunk 7]
```diff
@@ -62,12 +68,16 @@ def single_logout_service_url()
         val
       end
 
-      # setter
+      # Setter for the Single Logout Service URL.
+      # @param val [String].
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/settings.rb]: is `val` String an URL?

File: lib/onelogin/ruby-saml/settings.rb [hunk 8]
```diff
@@ -62,12 +68,16 @@ def single_logout_service_url()
...
+      #
       def single_logout_service_url=(val)
         @single_logout_service_url = val
       end
 
-      def single_logout_service_binding()
+      # @return [String] Single Logout Service Binding.
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/settings.rb]: is it possible to have a reference to available bindings?

File: lib/onelogin/ruby-saml/slo_logoutrequest.rb [hunk 9]
```diff
@@ -7,11 +7,19 @@
...
+    #
     class SloLogoutrequest < SamlMessage
       attr_reader :options
       attr_reader :request
       attr_reader :document
 
+      # Constructs the Logout Request. A Logout Request Object that is an extension of the SamlMessage class.
+      # @param request [String] A UUEncoded Logout Request from the IdP.
+      # @param options [Hash]   Some options for the logout request validation process like allow a clock drift when checking dates with :allowed_clock_drift
+      # @raise [ArgumentError]
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/slo_logoutrequest.rb]: in which cases is this `Exception` raised?

File: lib/onelogin/ruby-saml/slo_logoutrequest.rb [hunk 10]
```diff
@@ -19,15 +27,23 @@ def initialize(request, options = {})
...
+      # Validates the Logout Request with the default values (soft = true)
+      # @return [Boolean] TRUE if the Logout Request is valid
+      #
       def is_valid?
         validate
       end
 
+      # Validates the Logout Request (soft = false)
+      # @return [Boolean] TRUE if the Logout Request is valid
+      # @raise [ValidationError]
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/slo_logoutrequest.rb]: in which cases?

File: lib/onelogin/ruby-saml/slo_logoutrequest.rb [hunk 11]
```diff
@@ -52,10 +72,20 @@ def issuer
 
       private
 
+      # Validates the Logout Request
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the logout request is invalid or not)
+      # @return [Boolean] TRUE if the Logout Request is valid
+      # @raise [ValidationError] if soft == false and validation fails
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/slo_logoutrequest.rb]: +1 this description looks good example to follow for the above comments

File: lib/onelogin/ruby-saml/slo_logoutresponse.rb [hunk 12]
```diff
@@ -3,16 +3,31 @@
...
+      #
       def initialize
         @uuid = "_" + UUID.new.generate
       end
 
+      # Creates the Logout Response string.
+      # @param settings [OneLogin::RubySaml::Settings|nil] Toolkit settings
+      # @param request_id [String] The ID of the LogoutRequest sent by this SP to the IdP. That ID will be placed as the InResponseTo in the logout response
+      # @param logout_message [String] The Message to be placed as StatusMessage in the logout response
+      # @param params [Hash]   Some parameters to build the logout request
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/slo_logoutresponse.rb]: do you have any clues on what the `@option`s this `params` hash can accept?

File: lib/onelogin/ruby-saml/utils.rb [hunk 13]
```diff
@@ -1,42 +1,59 @@
...
+      # @param key [String] The original private key 
+      # @param heads [Boolean] If true, the formatted private key will include the
+      #                  "BEGIN PRIVATE KEY" or the "BEGIN RSA PRIVATE KEY" header and the footer.
+      # @return [String] The formatted certificate
+      #
       def self.format_private_key(key, heads=true)
-        key = key.delete("\n").delete("\r").delete("\x0D")
-        if key
+        if key && !key.empty?
+          key = key.delete("\n\r\x0D")
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/utils.rb]:
```ruby
key.delete!("\n\r\x0D")
```

File: lib/onelogin/ruby-saml/attribute_service.rb [hunk 14]
```diff
@@ -14,18 +19,29 @@ def configure(&block)
...
 
+      # Set a name 
+      # @param name [String] The service name
+      #
       def service_name(name)
         @name = name
       end
 
+      # Set an index 
+      # @param name [Integer] An index
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/attribute_service.rb]: `@param` should describe `index` instead

File: lib/onelogin/ruby-saml/attribute_service.rb [hunk 15]
```diff
@@ -14,18 +19,29 @@ def configure(&block)
...
+      #
       def service_index(index)
         @index = index
       end
-      
+
+      # Add an attribute
+      # @param name [Hash] Attribute for the AttributeService
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/attribute_service.rb]: `@param` should describe `options` and if possible, please provide `@option`s of the possible options that this method accepts

File: lib/onelogin/ruby-saml/attributes.rb [hunk 16]
```diff
@@ -1,91 +1,110 @@
...
 
-      # Get current status of backwards compatibility mode.
+      # @return [Boolean] Get current status of backwards compatibility mode.
+      #
       def self.single_value_compatibility
         @@single_value_compatibility
       end
 
       # Sets the backwards compatibility mode on/off.
+      # @param value [Boolean]
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/attributes.rb]: what will this `Boolean` represent?

File: lib/onelogin/ruby-saml/attributes.rb [hunk 17]
```diff
@@ -1,91 +1,110 @@
...
       # Test attribute presence by name
+      # @param name [String] The attribute name to be checked
+      #
       def include?(name)
         attributes.has_key?(canonize_name(name))
       end
       
       # Return first value for an attribute
+      # @param name [String] The attribute name
+      # return [String] The value (First occurrence)
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/attributes.rb]: should be `@return`

File: lib/onelogin/ruby-saml/attributes.rb [hunk 18]
```diff
@@ -1,91 +1,110 @@
...
+      #
       def single(name)
         attributes[canonize_name(name)].first if include?(name)
       end
 
       # Return all values for an attribute
+      # @param name [String] The attribute name
+      # return [Array] Values of the attribute
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/attributes.rb]: `@return`

File: lib/onelogin/ruby-saml/logoutresponse.rb [hunk 19]
```diff
@@ -3,23 +3,27 @@
...
+# Only supports SAML 2.0
 module OneLogin
   module RubySaml
+
+    # SAML2 Logout Response (SLO IdP initiated, Parser)
+    #
     class Logoutresponse < SamlMessage
-      # For API compability, this is mutable.
+
+      # OneLogin::RubySaml::Settings Toolkit settings
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/logoutresponse.rb]: awesome, let's make sure we have the same comment for other classes

File: lib/onelogin/ruby-saml/logoutresponse.rb [hunk 20]
```diff
@@ -3,23 +3,27 @@
...
-      #   :matches_request_id => REQUEST_ID
-      #
-      # It will validate that the logout response matches the ID of the request.
-      # You can also do this yourself through the in_response_to accessor.
+      # Constructs the Logout Response. A Logout Response Object that is an extension of the SamlMessage class.
+      # @param response  [String] A UUEncoded logout response from the IdP.
+      # @param settings  [OneLogin::RubySaml::Settings|nil] Toolkit settings
+      # @param options   [Hash] Extra parameters. 
+      #                    :matches_request_id It will validate that the logout response matches the ID of the request.
+      # @raise [ArgumentError]
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/logoutresponse.rb]: in which cases?

File: lib/onelogin/ruby-saml/saml_message.rb [hunk 21]
```diff
@@ -33,20 +45,20 @@ def valid_saml?(document, soft = true)
         end
       end
 
+      # Raise a ValidationError with the provided message
+      # @param message [String] Message of the exception
+      # @raise [ValidationError]
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: when will we expect a `ValidationError` ?

File: lib/onelogin/ruby-saml/slo_logoutresponse.rb [hunk 22]
```diff
@@ -25,6 +40,13 @@ def create(settings, request_id = nil, logout_message = nil, params = {})
         @logout_url = settings.idp_slo_target_url + response_params
       end
 
+      # Creates the Get parameters for the logout response.
+      # @param settings [OneLogin::RubySaml::Settings|nil] Toolkit settings
+      # @param request_id [String] The ID of the LogoutRequest sent by this SP to the IdP. That ID will be placed as the InResponseTo in the logout response
+      # @param logout_message [String] The Message to be placed as StatusMessage in the logout response
+      # @param params [Hash]  Some parameters to build the logout response
+      # @return [Hash] Parameters
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/slo_logoutresponse.rb]: what kind of `Parameters` ? is it possible we get more detail about the return value?

File: lib/onelogin/ruby-saml/attributes.rb (hunk 16 above)
Comment (pitbulk) [inline lib/onelogin/ruby-saml/attributes.rb]: The line describe the behavior: Turn the backwards compatibility mode on/off.

File: lib/onelogin/ruby-saml/saml_message.rb (hunk 21 above)
Comment (pitbulk) [inline lib/onelogin/ruby-saml/saml_message.rb]: This is the ValidationError call

File: lib/onelogin/ruby-saml/attribute_service.rb [hunk 23]
```diff
@@ -14,18 +19,29 @@ def configure(&block)
...
+      # @param name [String] The service name
+      #
       def service_name(name)
         @name = name
       end
 
+      # Set an index 
+      # @param name [Integer] An index
+      #
       def service_index(index)
```
Comment (daniel-g) [inline lib/onelogin/ruby-saml/attribute_service.rb]: why not service_index= ?

File: lib/onelogin/ruby-saml/attribute_service.rb [hunk 24]
```diff
@@ -14,18 +19,29 @@ def configure(&block)
...
+      # @return [Boolean] True if the AttributeService object has been initialized and set with the required values
+      #                   (has attributes and a name)
       def configured?
         @attributes.length > 0 && !@name.nil?
       end
 
+      # Set a name 
+      # @param name [String] The service name
+      #
       def service_name(name)
```
Comment (daniel-g) [inline lib/onelogin/ruby-saml/attribute_service.rb]: why not service_name= ?

Comment (luisvm) [inline lib/onelogin/ruby-saml/attribute_service.rb]: +1

File: lib/onelogin/ruby-saml/attributes.rb (hunk 17 above)
Comment (daniel-g) [inline lib/onelogin/ruby-saml/attributes.rb]: return [NilClass] If the value was not found

File: lib/onelogin/ruby-saml/saml_message.rb (hunk 21 above)
Comment (luisvm) [inline lib/onelogin/ruby-saml/saml_message.rb]: jesus, I feel like this wrapper is overkill, do we use this in many places?

File: lib/onelogin/ruby-saml/attributes.rb (hunk 18 above)
Comment (daniel-g) [inline lib/onelogin/ruby-saml/attributes.rb]: if it has to be always an array, wrap the response. Be careful:
```
1.8.7-head :001 > nil.to_a
 => []
1.8.7-head :002 > [].to_a
 => []
1.8.7-head :003 > {}.to_a
 => []
1.8.7-head :004 > {:a => 1, :b => 2}.to_a
 => [[:a, 1], [:b, 2]]
```

File: lib/onelogin/ruby-saml/attributes.rb (hunk 16 above)
Comment (luisvm) [inline lib/onelogin/ruby-saml/attributes.rb]: kk, we can dismiss this one

File: lib/onelogin/ruby-saml/response.rb [hunk 25]
```diff
@@ -49,6 +67,12 @@ def name_id
         end
       end
 
+      # Gets the SessionIndex from the AuthnStatement.
+      # Could be used to be stored in the local session in order
+      # to be used in a future Logout Request that the SP could
+      # send to the SP, to set what specific session must be deleted
```
Comment (luisvm) [inline lib/onelogin/ruby-saml/response.rb]: `... that the SP could send to the IdP ...`

File: lib/onelogin/ruby-saml/logoutresponse.rb [hunk 26]
```diff
@@ -30,16 +34,29 @@ def initialize(response, settings = nil, options = {})
         @document = XMLSecurity::SignedDocument.new(@response)
       end
 
+      # Aux function to validate the Logout Response (soft = false)
```
Comment (daniel-g) [inline lib/onelogin/ruby-saml/logoutresponse.rb]: I'd say 'Hard aux function to validate the Logout Response'

File: lib/onelogin/ruby-saml/attribute_service.rb (hunk 23 above)
Comment (pitbulk) [inline lib/onelogin/ruby-saml/attribute_service.rb]: I was only adding comments on this PR, the code on master is without the =.

Notice that the attribute is index, and we named the setter service_index.

File: lib/onelogin/ruby-saml/logoutresponse.rb (hunk 26 above)
Comment (pitbulk) [inline lib/onelogin/ruby-saml/logoutresponse.rb]: Ok, fixed similar occurrences

File: lib/onelogin/ruby-saml/attributes.rb (hunk 18 above)
Comment (pitbulk) [inline lib/onelogin/ruby-saml/attributes.rb]: PR for that is welcome since you guys are the Ruby Masters ;)

**Comment by luisvm** [comment]
:+1: lgmt

**Comment by daniel-g** [comment]
:+1:

**Comment by reedloden** [comment]
What are the security concerns around "Fix xpath injection on xml_security.rb"? Is that exploitable in any way? If so, should get a CVE.

I also don't see a test focused on that issue. Is there one?

**Comment by pitbulk** [comment]
Here:
https://github.com/onelogin/ruby-saml/pull/225/files#diff-661b9d9743a3ff77661f224c6191165cR259
lib/xml_security.rb  L259

It is similar bug than https://github.com/onelogin/ruby-saml/pull/183
http://osvdb.org/show/osvdb/117903

**Comment by reedloden** [comment]
Sounds like it needs a CVE then. I'll put in a request for one.

**Comment by pitbulk** [comment]
Ok, thanks @reedloden

**Comment by reedloden** [comment]
Requested in http://seclists.org/oss-sec/2015/q3/74

**Comment by reedloden** [comment]
Also, I went ahead and opened https://github.com/onelogin/ruby-saml/issues/252 to figure out a better process for handling security issues for the future.

**Comment by maikypedia** [comment]
The CVE assigned is [CVE-2015-20108](CVE-2015-20108)

**Pull request merged**